rfcs/20190501-tf-tensor-cord.md

```diff
@@ -0,0 +1,244 @@
+standpoint; e.g. it adds a `CHECK` to the `Tensor::Tensor(dtype, shape)`
+constructor so users don't accidentally create a `DT_CORD` tensor without
+a releaser.
+
+| Alternatives | TensorCord DT_VARIANT | DT_CORD |
+:------------- |:--------------------- |:------- |
+| Separate releasers per element | Yes | No |
+| Overhead | Higher (each Tensor element keeps a reference; Variant & RunVariantDtor overhead is more costly) | Lower (can have onereleaser per tensor) |
+| Intrusiveness | Lower (use DT_VARIANT) | Higher (add new TF type) |
+| Flexibility | High (elements can point to locations backed by different owners) | Lower (all elements must be backed by data whose lifetime depends a shared set of releasers)
+
+## Detailed Design
+
+### Public C++ API
+
+The TensorCord object constructor and Append methods accept a string_view and
+either a Tensor pointer or releaser callback.  Its underlying string views can
+be iterated over or a string can be constructed via explicit cast:
+
+```c++
+class TensorCord {
+public:
+  typedef void (*Releaser)(void*);
+
+  // At final destruction, releaser will be called as releaser(memory).
+  // To create a releaser for a std::function that captures objects, use:
+  //
+  //  template <typename T>
+  //  TensorCord::Releaser CreateThunkFor(const T& fn) {
+  //    return [](void* ptr) { (*static_cast<T*>(ptr))(); };
+  // }
+  //
+  // auto fn = [&]() { ... };
+  // auto releaser = CreateThunkFor(fn);
+  // auto tc = TensorCord(view, releaser, &fn);
+  //
+  // Remember that in this case, fn needs to outlast the TensorCord.
+  //
+  // Creates a TensorCord from `view`, with memory releaser `releaser` and releaser
+  // arg `memory`.
+  explicit TensorCord(absl::string_view view, Releaser releaser,
+                                  void* memory = nullptr);
+
+  // Creates a TensorCord from `view`, with memory backed by `tensor`.  If it `view`
+  // is small enough, no reference is created on `tensor`; instead the memory is
+  // stored inline.
+  explicit TensorCord(absl::string_view view, Tensor* tensor);
+  explicit TensorCord(absl::string_view view, RefCounted* ref_counted);
+
+  void Append(const TensorCord& other);
+  void Append(absl::string_view view, CordRep::Releaser releaser,
+                        void* memory = nullptr);
+  void Append(absl::string_view view, Tensor* tensor);
+  void Append(absl::string_view view, RefCounted* ref_counted);
+
+  size_t size() const;
+  bool empty() const;
+
+  explicit operator string() const;
+
+  // Usage example:
+  //  for (absl::string_view s : cord.Chunks()) { ... }
+  ChunkRange Chunks() const;
+  ChunkIterator chunk_begin() const;
+  ChunkIterator chunk_end() const;
+
+  // Copy and move constructor, copy and move assignment operators.
+  // And all the associated Variant-stored object methods
+  // (Encode, Decode, DebugString, etc).
+};
+```
+
+### Ops and Op extensions supporting TensorCord:
+
+The following ops would be extended to support TensorCord:
```

Comment:
Could you give an idea of what it would take to extend an Op to support TensorCord?
E.g., is extending `join` as simple as templatizing the Op for both string and TensorCord?

Reply:
I'll update the doc with a corresponding code example.  It's pretty similar to what you said: you first templatize the Op for both DT_STRING and DT_VARIANT.  Then in the kernel you check the input dtype; if it's Variant, you access an Eigen tensor of Variant objects.  From each variant in the Tensor, you access the underlying TensorCord via `get<TensorCord>()`.  If the pointer is nullptr, error out (because the stored object is not a TensorCord); otherwise access it and perform the join.

For the specific case case of the join op, you may actually prefer to emit a new TensorCord Variant Tensor containing the joined strings as sub-Cords (converting to string would perform the concatenation).